rules/rbe_repo.bzl

```diff
@@ -0,0 +1,597 @@
+                --extra_toolchains=@rbe_default//platforms:cc-toolchain \
+
+    Note running bazel clean --expunge_async, or otherwise modifying attrs or
+    env variables used by this rule will trigger it to re-execute. Running this
+    repo rule takes some time as it needs to pull a container, run it, and then
+    run some commands inside. We recommend you use output_base and check in the produced
+    files so you dont need to run this rule with every clean build.
+
+The {bazel_version} above corresponds to the version of bazel installed locally.
+Note you can override this version and pass an optional rc# if desired.
+Running this rule with a non release version (e.g., built from source) will not work.
+If running with bazel built from source you must pass a bazel_version and bazel_rc
+to rbe_autoconfig. Also, note the bazel_version bazel_rc must be published in 
+https://releases.bazel.build/...
+
+Note this is a very not hermetic repository rule that can actually change the
+contents of your project sources. While this is generally not recommended by
+Bazel, its the only reasonable way to get a rule that can produce valid
+toolchains / platforms that need to be made available to Bazel before execution
+of any build actions.
+
+Note: this rule expects the following utilities to be installed and available on
+the PATH:
+  - docker
+  - gunzip
+  - tar
+  - sha256sum
+  - bash utilities (e.g., cp, mv, rm, etc)
+
+Known issues:
+  This rule cannot be executed inside a docker container.
+"""
+
+load("@io_bazel_rules_docker//container:pull.bzl", _pull = "pull")
+load(
+    "@bazel_toolchains//rules:version_check.bzl",
+    "check_bazel_version",
+    "extract_version_number",
+    "parse_rc",
+)
+load(
+    "@bazel_toolchains//rules:toolchain_containers.bzl",
+    "RBE_UBUNTU16_04_LATEST",
+    "public_rbe_ubuntu16_04_sha256s",
+)
+
+# External folder is set to be deprecated, lets keep it here for easy
+# refactoring
+# https://github.com/bazelbuild/bazel/issues/1262
+_EXTERNAL_FOLDER_PREFIX = "external/"
+RBE_CONFIG_DIR = "rbe_config_cc"
+PLATFORM_DIR = "platforms"
+BAZEL_CONFIG_DIR = "/bazel-config"
+PROJECT_REPO_DIR = "project_src"
+OUTPUT_DIR = BAZEL_CONFIG_DIR + "/autoconf_out"
+REPO_DIR = BAZEL_CONFIG_DIR + "/" + PROJECT_REPO_DIR
+VERBOSE = False
+RBE_AUTOCONF_ROOT = "RBE_AUTOCONF_ROOT"
+CONFIG_REPOS = ["local_config_cc"]
+
+def _impl(ctx):
+    """Core implementation of _rbe_autoconfig repository rule."""
+    project_root = ctx.os.environ.get(RBE_AUTOCONF_ROOT, None)
+    use_default_project = False
+    if not project_root:
+        if ctx.attr.output_base != "":
+            fail(("%s env variable must be set for rbe_autoconfig" +
+                  " to function properly when output_base is set") % RBE_AUTOCONF_ROOT)
+
+        # Try to use the default project
+        project_root = ctx.path(".").dirname.get_child("bazel_toolchains").get_child("rules").get_child("cc-sample-project")
```

Comment:
Don't fully understand how this works? Why would "bazel_toolchains" be a path under the project root? Does ctx.path(".").dirname.get_child automatically resolve external repositories somehow?

Reply:
yes, '.' resolves to the root of this external repository, then you can traverse up and down to find the right target in the bazel_toolchains repo. This is the kind of black magic you are allowed to do in repository rules that you can never do in regular build rules.